### ISSUE TITLE
Implement SileroVAD Wrapper

### ISSUE THREAD

**Issue by ArthurCRodrigues** [inference4j/inference4j#19]
# Silero VAD Implementation for inference4j

## Overview

This issue details the implementation of a wrapper for [Silero VAD](https://github.com/snakers4/silero-vad) (Voice Activity Detection) to inference4j, enabling speech segment detection in audio files. 

Silero VAD is a lightweight, accurate voice activity detector that identifies when speech occurs in audio streams.

## New Files

- `inference4j-models/src/main/java/io/github/inference4j/audio/SileroVAD.java` - Main model wrapper
- `inference4j-models/src/main/java/io/github/inference4j/audio/VoiceSegment.java` - Output record for detected segments
- `inference4j-models/src/test/java/io/github/inference4j/audio/SileroVADTest.java` - Unit tests
- `inference4j-examples/src/main/java/io/github/inference4j/examples/VoiceActivityDetectionExample.java` - Usage example

### Basic Usage

```java
try (SileroVAD vad = SileroVAD.fromPretrained("models/silero-vad")) {
    List<VoiceSegment> segments = vad.detect(Path.of("audio.wav"));
    for (VoiceSegment segment : segments) {
        System.out.printf("Speech: %.2fs - %.2fs (confidence: %.2f)%n",
            segment.start(), segment.end(), segment.confidence());
    }
}
```

### Custom Configuration

```java
try (SileroVAD vad = SileroVAD.builder()
        .session(InferenceSession.create(modelPath))
        .threshold(0.6f)           // Speech probability threshold
        .minSpeechDuration(0.25f)  // Minimum speech segment duration (seconds)
        .minSilenceDuration(0.1f)  // Minimum silence to end a segment (seconds)
        .build()) {
    List<VoiceSegment> segments = vad.detect(audioSamples, 16000);
}
```

### Running the Example

```bash
./gradlew :inference4j-examples:run -PmainClass=io.github.inference4j.examples.VoiceActivityDetectionExample
```

### Model Architecture

Silero VAD v5 is a recurrent neural network that processes audio in small chunks (512 samples at 16kHz = 32ms per frame). It maintains hidden state between frames using LSTM layers, allowing it to track speech patterns over time.

**Model Inputs:**
| Name | Shape | Type | Description |
|------|-------|------|-------------|
| `input` | `[batch, context + window]` | float32 | Audio samples with context prepended |
| `state` | `[2, batch, 128]` | float32 | LSTM hidden state (h, c combined) |
| `sr` | `[]` (scalar) | int64 | Sample rate (8000 or 16000) |

**Model Outputs:**
| Name | Shape | Type | Description |
|------|-------|------|-------------|
| `output` | `[batch, 1]` | float32 | Speech probability (0.0 - 1.0) |
| `stateN` | `[2, batch, 128]` | float32 | Updated LSTM state for next frame |

### Key Implementation Challenge: Context Buffer

The most significant challenge encountered was understanding that **Silero VAD requires a context buffer to be prepended to each audio chunk**.

#### The Problem

Initial implementation passed only the 512-sample window directly to the model:

```java
// INCORRECT - Missing context
float[] window = new float[512];
System.arraycopy(samples, start, window, 0, 512);
Tensor inputTensor = Tensor.fromFloats(window, new long[]{1, 512});
```

This resulted in the model outputting extremely low probabilities (max ~0.003) even for audio files with clear speech, making it impossible to detect any voice activity.

#### Root Cause Analysis

After investigating the [official Silero VAD Python implementation](https://github.com/snakers4/silero-vad/blob/master/src/silero_vad/utils_vad.py), we discovered the `OnnxWrapper.__call__` method:

```python
context_size = 64 if sr == 16000 else 32
x = torch.cat([self._context, x], dim=1)  # Prepend context!
```

The model expects:
- **64 samples of context** prepended to each 512-sample window (at 16kHz)
- **32 samples of context** prepended to each 256-sample window (at 8kHz)
- Context is initialized with zeros for the first frame
- Context is updated with the last N samples from each processed window

#### The Solution

The corrected implementation maintains a context buffer that persists across frames:

```java
private float[] runInference(float[] samples) {
    int numFrames = (int) Math.ceil((double) samples.length / windowSizeSamples);
    float[] probabilities = new float[numFrames];

    // Initialize LSTM state
    float[] state = new float[STATE_LAYERS * 1 * HIDDEN_SIZE];

    // Initialize context buffer with zeros (critical for Silero VAD v5)
    float[] context = new float[contextSize];  // 64 samples at 16kHz

    for (int frame = 0; frame < numFrames; frame++) {
        // Extract current window
        float[] window = new float[windowSizeSamples];
        System.arraycopy(samples, frame * windowSizeSamples, window, 0, ...);

        // Create input WITH context prepended (576 samples total)
        float[] inputWithContext = new float[contextSize + windowSizeSamples];
        System.arraycopy(context, 0, inputWithContext, 0, contextSize);
        System.arraycopy(window, 0, inputWithContext, contextSize, windowSizeSamples);

        // Run inference with context-augmented input
        Tensor inputTensor = Tensor.fromFloats(inputWithContext, 
            new long[]{1, contextSize + windowSizeSamples});
        
        // ... run model and get probability ...

        // Update context for next frame (last 64 samples of current window)
        System.arraycopy(window, windowSizeSamples - contextSize, context, 0, contextSize);
    }
    return probabilities;
}
```

#### Results

| Metric | Before Fix | After Fix |
|--------|------------|-----------|
| Max probability | 0.0037 | 1.0000 |
| Speech frames detected | 0 | 22+ per file |
| Segments detected | 0 | 5 (on test file) |

### Audio Processing Pipeline

The implementation reuses existing audio preprocessing from the `inference4j-preprocessing` module:

1. **Audio Loading** (`AudioLoader.load()`) - Reads WAV files and extracts raw samples
2. **Resampling** (`AudioProcessor.resample()`) - Converts to 16kHz if needed
3. **Normalization** - Audio is expected in [-1.0, 1.0] float range

### Segment Extraction Algorithm

After obtaining per-frame probabilities, segments are extracted with hysteresis to avoid fragmentation:

```
For each frame:
  if probability >= threshold:
    if not in_speech:
      start new segment
    reset silence counter
  else:
    if in_speech:
      increment silence counter
      if silence_counter >= min_silence_frames:
        end segment (if duration >= min_speech_duration)
```

This approach:
- Requires configurable minimum speech duration (default: 250ms)
- Allows brief silence within speech (default: up to 100ms)
- Filters out noise spikes that briefly exceed threshold

## Configuration Options

| Parameter | Default | Description |
|-----------|---------|-------------|
| `threshold` | 0.5 | Speech probability threshold |
| `minSpeechDuration` | 0.25s | Minimum segment length to report |
| `minSilenceDuration` | 0.1s | Silence duration to end a segment |
| `sampleRate` | 16000 | Target sample rate (8000 or 16000) |
| `windowSizeSamples` | 512 | Samples per frame (512 at 16kHz, 256 at 8kHz) |

## Testing

Unit tests cover:
- Segment extraction logic (without ONNX runtime)
- Single and multiple speech segments
- Short speech filtering
- Speech at end of audio
- Confidence calculation
- Builder validation

Integration testing requires the Silero VAD ONNX model file.

## Model Setup

Download the Silero VAD ONNX model:

```bash
mkdir -p inference4j-examples/models/silero-vad
# Download from https://github.com/snakers4/silero-vad/tree/master/src/silero_vad/data
# Place model.onnx in the silero-vad directory
```

## Future Improvements

- [ ] Streaming/real-time VAD support (process audio chunks as they arrive)
- [ ] 8kHz sample rate support (telephone audio)
- [ ] Batch processing for multiple audio files
- [ ] Integration with Wav2Vec2 for VAD-assisted transcription
